### DISCUSSION TITLE
Questions about archiving SS exes/tagging source code?

### DISCUSSION THREAD

**Discussion by k-doering-NOAA** [nmfs-stock-synthesis/stock-synthesis#425] [Team Posts]
Hi all - Chantel and I are having a discussion with Owen, which has brought up some questions for me about archiving SS versions. We have been using Google drive to do this (https://drive.google.com/drive/u/0/folders/0Bz1UsDoLaOMLYjE5OTdiY2QtM2ZlMi00NzExLTk4M2MtYzI3YTI2YWJiMDUz) . However, Owen needs a Mac version of 3.30.01.00 to run projections, but for a lot of versions of SS around this time (~3.30.01 to 3.30.10), there aren't any mac or linux versions archived. The code also isn't [tagged](https://github.com/nmfs-stock-synthesis/stock-synthesis/tags) , although I did add v3.30.01 and v3.30.02 tags this morning (we can edit these if I didn't tag them in the right place). The [release messages on the forums](https://vlab.ncep.noaa.gov/web/stock-synthesis/public-forums/-/message_boards/category/988476) indicate to me that there wasn't always a release version (like 3.30.xx.00) between 3.30.02 and 3.30.10. I'm sure all of you know better than me, as this was before my time! Some questions:

1. Is it worth the effort to try to tag the code base between 3.30.02 and 3.30.10? Maybe tags for which there was a corresponding release message? This is pretty simple to do, it is just a matter of identifying which version is which from the commit messages.
2. Should we try to recompile mac and linux versions that are missing for old versions of SS and add them to the archive? This is a lot more effort and may take some time, as an older version of admb will also need to be used. I just tried compiling the 11.2 admb source code on the mac mini, but didn't have any luck, so I'm not sure this will even pan out (will try github actions, though...). I think this is rarely requested, but it does worry me a little bit that folks that have used an older version for an assessment may not be able to use the exact same version if they need to run projections, for instance. 

Or, maybe do we leave it up to the user to figure out how to use the windows exe? It's a lot of effort, but someone could [install a windows virtual machine on their mac](https://www.howtogeek.com/657464/how-to-install-a-windows-10-virtualbox-vm-on-macos/) . I had to do this for a pop dy class I took in 2014 in order to use tools from the old NMFS fisheries toolbox, so maybe it is not crazy to suggest this as a solution?

**Comment by Rick-Methot-NOAA** [comment]
Thanks Kathryn for laying this out clearly.
I support creating a tagged archive for earlier versions of the tpls
On my local machine, I have archive version of tpl and exe for 3.30.01, then continuous beginning with 3.30.10.  But nothing for 3.30.02 thru 3.30.09.  Old versions of the change log should document what was happening and could help recreate.
I do not support creating an archive back in time for linux and mac as the cost/benefit does not seem favorable.

**Comment by iantaylor-NOAA** [comment]
I think a tagged version of the archive would be adequate too. For some
users, that might be adequate as they could download the source associated
with that tag, install ADMB on their Mac or whatever machine, and compile
whatever version they need. However, there have been some updates to the
make files and the compiling process.
I think trying to come up with a backwards compatible set of compiling
instructions would be challenging, so it might be better to simply have
folks reach out to the SS development team if they need an executable which
isn't present in the archive, and we could figure out how to compile it.

**Comment by k-doering-NOAA** [comment]
Ok Rick and Ian, that makes sense to me - the cost/benefit of tagging the repository also seems worth it to me, but I agree that trying to recreate the archive does not. 
Luckily, the google drive does have a good archive of windows exes between 3.30.02 and 3.30.09 (looks like thanks to Chantel and Ian :) ), so I think we are ok in terms of having the windows versions we need.

I am giving a try of using github actions to compile an older mac version (the one Owen requested), but if it doesn't work, then I won't put anymore effort into this.

**Comment by kellijohnson-NOAA** [comment]
I think that the amount of effort spent compiling old versions depends on the question that is being answered with the older version. My guess is that Owen wants to provide advice to managers and that we should provide support for getting his files up to speed with the latest version rather than worrying about compiling old versions.

**Comment by k-doering-NOAA** [comment]
Hmm, I guess that is a question is up to you all in PEP (or in general whatever scientists are providing advice to managers), but that is a good point. Perhaps the real solution is just upgrading the model. Thanks, Kelli!

**Comment by chantelwetzel-noaa** [comment]
To clarify, there are situations where updating the model version would not
be an option.  I think the situation that Owen is in for the catch only
updates is one that he would not be allowed to update the model if an
updated model does not match the existing model results exactly.
Alternatively, in new assessments and updates one would have more freedom
to update to a newer version of SS.

Chantel Wetzel, PhD

*Pronouns: she/her (what does that mean? <https://www.mypronouns.org/>)*

*Fishery Resource Analysis and Monitoring DivisionNorthwest Fisheries
Science Center*
*National Marine Fisheries Service*
*National Oceanic and Atmospheric Administration*
*Office: 206.302.1753 *

**Comment by k-doering-NOAA** [comment]
thanks, Chantel. Owen confirmed that the version can't change. I didn't have any luck compiling ADMB 11.2 on the mac mini or gh actions, so I'm wondering if it would be ok in this instance to try compiling the old source code with a newer version of admb so Owen can check if the results are the same as with the windows exe with the correct admb version?(perhaps try 12.0 to start). I think ultimately, this is so students with macs can follow along and contribute, but others will also be using the windows exe.

@Rick-Methot-NOAA , does it seem ok to compile with a newer version of ADMB in this case?

**Comment by Rick-Methot-NOAA** [comment]
OK to compile old code with new ADMB.  We do not have any examples of ADMB
dependency that I can recall.
Rick

**Comment by k-doering-NOAA** [comment]
Yes, since we've updated to 12.2 and 12.3, our test sets hasn't changed, which makes me think it is unlikely that which ADMB version really matters.
